core/server/master/src/main/java/alluxio/master/file/AccessTimeUpdater.java

```diff
@@ -0,0 +1,176 @@
+import alluxio.master.file.meta.InodeTree.LockPattern;
+import alluxio.master.file.meta.LockedInodePath;
+import alluxio.master.journal.JournalContext;
+import alluxio.master.journal.JournalSystem;
+import alluxio.master.journal.sink.JournalSink;
+import alluxio.proto.journal.File.UpdateInodeEntry;
+import alluxio.proto.journal.Journal;
+import alluxio.resource.LockResource;
+import alluxio.util.ThreadFactoryUtils;
+import alluxio.util.ThreadUtils;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Iterator;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * This class represents the executor for periodic inode ttl check.
+ */
+@NotThreadSafe
+final class AccessTimeUpdater implements JournalSink {
+  private static final Logger LOG = LoggerFactory.getLogger(AccessTimeUpdater.class);
+
+  private final long mFlushInterval;
+  private final long mUpdatePrecision;
+  private final long mShutdownTimeout;
+
+  private final FileSystemMaster mFileSystemMaster;
+  private final InodeTree mInodeTree;
+
+  /** Keep track of all inodes that need access time update. */
+  private ConcurrentHashSet<Long> mAccessTimeUpdates;
+  private ScheduledExecutorService mExecutorService = null;
+  private AtomicBoolean mUpdateScheduled = new AtomicBoolean();
+
+  /**
+   * Constructs a new {@link AccessTimeUpdater}.
+   */
+  public AccessTimeUpdater(FileSystemMaster fileSystemMaster, InodeTree inodeTree,
+      JournalSystem journalSystem) {
+    this(fileSystemMaster, inodeTree, journalSystem,
+        ServerConfiguration.getMs(PropertyKey.MASTER_FILE_ACCESS_TIME_JOURNAL_FLUSH_INTERVAL),
+        ServerConfiguration.getMs(PropertyKey.MASTER_FILE_ACCESS_TIME_UPDATE_PRECISION),
+        ServerConfiguration.getMs(PropertyKey.MASTER_FILE_ACCESS_TIME_UPDATER_SHUTDOWN_TIMEOUT));
+  }
+
+  /**
+   * Constructs a new {@link AccessTimeUpdater} with time configurations.
+   */
+  @VisibleForTesting
+  public AccessTimeUpdater(FileSystemMaster fileSystemMaster, InodeTree inodeTree,
+      JournalSystem journalSystem, long flushInterval, long updatePrecision, long shutdownTimeout) {
+    mFileSystemMaster = fileSystemMaster;
+    mInodeTree = inodeTree;
+    mAccessTimeUpdates = new ConcurrentHashSet<>();
+    mFlushInterval = flushInterval;
+    mUpdatePrecision = updatePrecision;
+    mShutdownTimeout = shutdownTimeout;
+    journalSystem.addJournalSink(mFileSystemMaster, this);
+  }
+
+  public void start() {
+    start(mFlushInterval > 0 ? Executors.newSingleThreadScheduledExecutor(
+        ThreadFactoryUtils.build("AccessTimeUpdater-%d", true)) : null);
+  }
+
+  @VisibleForTesting
+  public void start(ScheduledExecutorService executorService) {
+    mExecutorService = executorService;
```

Comment:
what if `mExecutorService` was already  not `null` from a previous `start`?

Reply:
A previous `mExecutorService` should have been already shutdown by a previous `stop()`.

Reply:
`AccessTimeUpdater.stop()` is called on  `FileSystemMaster.stop()`. If that is not called before another start, the master will certainly be in a bad state. But just to be defensive, I will update it to try stop before starting a new executor, if needed.